File: chor-extraction/src/main/java/epp/ChoreographyVisitor.java
package epp;

import antlr4.ChoreographyBaseVisitor;
import antlr4.ChoreographyParser.*;
import ast.cc.interfaces.CCNode;
import ast.cc.interfaces.Choreography;
import ast.cc.nodes.*;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;

import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;

public class ChoreographyVisitor extends ChoreographyBaseVisitor<CCNode> {

    private HashSet<String> processes;

    public HashSet<String> getProcesses() {
        return processes;
    }

    public ChoreographyVisitor() {
        processes = new HashSet<>();
    }

    public CCNode getCCAST(ParseTree parseTree){
        return this.visit(parseTree);
    }

    @Override
    public CCNode visitCommunication(CommunicationContext ctx) {
        String sender = ctx.process(0).getText();
        String receiver = ctx.process(1).getText();
        String expression = ctx.expression().getText();

        CCNode continuation = visit(ctx.choreography());

        processes.add(sender);
        processes.add(receiver);

        return new Communication(sender,receiver,expression, continuation);
    }

    @Override
    public CCNode visitSelection(SelectionContext ctx) {

        String sender = ctx.process(0).getText();
        String receiver = ctx.process(1).getText();
        String expression = ctx.label().getText();

        CCNode continuation = visit(ctx.choreography());

        processes.add(sender);
        processes.add(receiver);

        return new Selection(sender,receiver,expression, continuation);
    }

    @Override
    public CCNode visitCondition(ConditionContext ctx) {
        String process = ctx.process().getText();
        String expression = ctx.expression().getText();
        CCNode thenChoreography = visit(ctx.choreography(0));
        CCNode elseChoreography = visit(ctx.choreography(1));

        processes.add(process);

        return new Condition(process, expression, thenChoreography, elseChoreography);
    }

    @Override
    public CCNode visitProgram(ProgramContext ctx) {
        List<CCNode> procedures = ctx.procedureDefinition().stream().map(this::visit).collect(Collectors.toList());
        return new Program((Choreography) visit(ctx.main()), procedures);
    }

    @Override
    public CCNode visitProcedureDefinition(ProcedureDefinitionContext ctx) {
        return new ProcedureDefinition(ctx.procedure().getText(), visit(ctx.choreography()), processes);
    }

    @Override
    public CCNode visitMain(MainContext ctx) {
        return visit(ctx.choreography());
    }

    @Override public CCNode visitProcedureInvocation(ProcedureInvocationContext ctx) {
        String procedureName = ctx.procedure().getText();
        return new ProcedureInvocation(procedureName, processes);
    }

    @Override
    public CCNode visitTerminal(TerminalNode node) {
        return new Termination();
    }
}

File: pe/src/main/kotlin/np/ChoreographyVisitor.kt
package np

import antlrgen.ChoreographyBaseVisitor
import antlrgen.ChoreographyParser
import ast.cc.interfaces.CCNode
import ast.cc.interfaces.ChoreographyBody
import ast.cc.nodes.*

/**
 * parse context received from antlr to choreography
 */
class ChoreographyVisitor : ChoreographyBaseVisitor<CCNode>() {
    override fun visitProgram(ctx: ChoreographyParser.ProgramContext): CCNode {
        val choreographies = ArrayList<Choreography?>()

        ctx.choreography().forEach { chor -> choreographies.add(visitChoreography(chor)) }

        return Program(choreographies = choreographies, statistics = ArrayList())
    }

    override fun visitChoreography(ctx: ChoreographyParser.ChoreographyContext): Choreography {
        val main = visitBehaviour(ctx.main().behaviour())
        val procedures = ArrayList<ProcedureDefinition>()
        ctx.procedureDefinition().forEach { procedure -> procedures.add(visitProcedureDefinition(procedure)) }

        return Choreography(main = main, procedures = procedures, processes = emptySet())
    }

    override fun visitBehaviour(ctx: ChoreographyParser.BehaviourContext): ChoreographyBody {
        return when {
            !ctx.condition().isEmpty -> visitCondition(ctx.condition())
            !ctx.interaction().isEmpty -> visitInteraction(ctx.interaction())
            !ctx.procedureInvocation().isEmpty -> visitProcedureInvocation(ctx.procedureInvocation())
            else -> return Termination()
        }
    }


    override fun visitProcedureDefinition(ctx: ChoreographyParser.ProcedureDefinitionContext): ProcedureDefinition {
        return ProcedureDefinition(name = ctx.procedure().text, body = visitBehaviour(ctx.behaviour()), usedProcesses = hashSetOf())
        TODO("used processes should be somehow counted")
    }

    override fun visitCondition(ctx: ChoreographyParser.ConditionContext): Condition {
        return Condition(process = ctx.process().text, expression = ctx.expression().text, thenChoreography = visitBehaviour(ctx.behaviour(0)), elseChoreography = visitBehaviour(ctx.behaviour(1)))
    }

    override fun visitInteraction(ctx: ChoreographyParser.InteractionContext): ChoreographyBody {
        return when {
            !ctx.communication().isEmpty -> {
                val communication = ctx.communication()
                CommunicationSelection(eta = Communication(sender = communication.process(0).text, receiver = communication.process(1).text, expression = communication.expression().text), continuation = visitBehaviour(communication.behaviour()))
            }
            !ctx.selection().isEmpty -> {
                val selection = ctx.selection()
                CommunicationSelection(eta = Selection(sender = selection.process(0).text, receiver = selection.process(1).text, label = selection.expression().text), continuation = visitBehaviour(selection.behaviour()))
            }
            else -> throw Exception()
        }
    }

    override fun visitProcedureInvocation(ctx: ChoreographyParser.ProcedureInvocationContext): ProcedureInvocation {
        return ProcedureInvocation(procedure = ctx.procedure().text, processes = hashSetOf())
    }
}